docs/DesignDocs/SGX2-EDMM.md

```diff
@@ -0,0 +1,105 @@
+# SGX2 EDMM in Open Enclave SDK
+
+Intel SGX2 on Icelake CPUs supports enclave dynamic memory management (EDMM),
+which many applications of the Open Enclave SDK (OE) will want to take advantage
+of. This document outlines the strategy for making these features available to
+applications.
+
+The key mechanism to support the dynamic addition of memory pages is the
+`ENCLS[EAUG]` instruction, which allocates and zeroes a new memory page via
+communication with the SGX driver. Before the enclave uses the new page, it
+acknowledges its addition by issuing an `ENCLU[EACCEPT]` instruction for the
+respective page address. Further, the `ENCLS[EMODPR]` and `ENCLU[EMODPE]`
+instructions restrict or extend page permissions, respectively. Page type
+changes (including page removal/trimming) are performed by the `ENCLS[EMODT]`
+instruction.
+
+## Heap allocation
+
+We propose to add a new OCALL to OE enclaves which requests new memory pages
```

Comment:
internal or exposed to apps?

Reply:
I think we should expose it too. We need a default (internal) implementation, but we also need to expose it to allow applications to implement their own strategies. This could conceivably be a part of the pluggable allocator API, but I think that would just end up being an exposed wrapper for an internal-only OCALL.